src/signature-schemes/BN254SignatureScheme.sol

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8;
+
+import {BLS} from "src/libraries/BLS.sol";
+
+import {BytesLib} from "src/libraries/BytesLib.sol";
+
+import {ISignatureScheme} from "src/interfaces/ISignatureScheme.sol";
+
+/// @title BN254SignatureScheme contract
+/// @author Randamu
+/// @notice A contract that implements a BN254 signature scheme
+contract BN254SignatureScheme is ISignatureScheme {
+    using BytesLib for bytes32;
+
+    /// @notice Enum to represent the contract type
+    enum ContractType {
+        Bridge,
+        Upgrade
+    }
+
+    /// @notice Links public keys of threshold network statically to signature scheme contracts and remove from constructor of sender contracts. Admin cannot update, simply use new scheme id.
+    BLS.PointG2 private publicKey;
+
+    /// @notice Signature scheme identifier
+    string public constant SCHEME_ID = "BN254";
+
+    /// @notice Domain separation tag for the BLS signature scheme
+    bytes public DST;
+
+    /// @notice Custom error for invalid contract type in the constructor.
+    /// @notice Should either be "bridge" or "upgrade"
+    error InvalidContractType();
+
+    /// @notice Constructor for the BN254SignatureScheme contract.
+    /// @param contractType The type of contract (0 for Bridge, 1 for Upgrade).
+    constructor(bytes memory publicKeyBytes, ContractType contractType) {
+        // Validate the contract type
+        if (contractType != ContractType.Bridge && contractType != ContractType.Upgrade) {
+            revert InvalidContractType();
+        }
+
+        publicKey = BLS.g2Unmarshal(publicKeyBytes);
+
+        // Set the DST based on the contract type
+        string memory typeString = contractType == ContractType.Bridge ? "bridge" : "upgrade";
+        DST = abi.encodePacked(
+            "dcipher-", typeString, "-v01-BN254G1_XMD:KECCAK-256_SVDW_RO_", bytes32(block.chainid).toHexString(), "_"
+        );
```

Comment:
The application name, "dcipher", should probably be an argument as well.
I also don't think it's necessary to limit ourselves to just 2 ContractType, could just be part of the application prefix when deploying.